src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -135,39 +134,63 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
     currentEventsByTag(tag, offset.value)
 
-  private def currentJournalEventsByTag(tag: String, offset: Long, max: Long): Source[(PersistentRepr, Set[String], JournalRow), NotUsed] = {
+  private def currentJournalEventsByTag(tag: String, offset: Long, max: Long, latestOrdering: MaxOrderingId): Source[EventEnvelope, NotUsed] = {
+    if (latestOrdering.maxOrdering < offset) Source.empty
+    else {
+      readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
+        .mapAsync(1)(Future.fromTry)
+        .mapConcat { case (repr, _, row) =>
+          adaptEvents(repr).map(r => EventEnvelope(Sequence(row.ordering), r.persistenceId, r.sequenceNr, r.payload))
+        }
+    }
+  }
+
+  /**
+    * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+    *                             query has been executed which might return an event with this offset (or a higher offset).
+    *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+    *                             will be returned completely.
+    */
+  private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
     import akka.pattern.ask
-    implicit val askTimeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    Source.fromFuture(orderingActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId])
-      .flatMapConcat { latestOrdering =>
-        if (latestOrdering.maxOrdering < offset) Source.empty
-        else readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
-      }
-      .mapAsync(1)(Future.fromTry)
+    implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
+    /* We unfold with 3 parameters:
+     * - The first parameter (from) determines the starting offset (i.e. the minimum value) to query from.
+     * - The second parameter (completeImmediately) signifies that we have queried events until at least
+     *   `terminateAfterOffset`, therefore the stream no longer needs to continue.
+     * - The third parameter (retrieveImmediately) is a boolean which determines whether the database query should be
+     *   executed immediately, or after a delay.
+     */
+    Source.unfoldAsync[(Long, Boolean, Boolean), Seq[EventEnvelope]]((offset, false, false)) {
+      case (from: Long, completeImmediately: Boolean, retrieveImmediately: Boolean) =>
+        def retrieveNextBatch(): Future[Some[((Long, Boolean, Boolean), Seq[EventEnvelope])]] = {
+          for {
+            queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
+            xs <- currentJournalEventsByTag(tag, from, readJournalConfig.maxBufferSize, queryUntil).runWith(Sink.seq)
+          } yield {
+            val isFullBatch = xs.size == readJournalConfig.maxBufferSize
```

Comment:
@WellingR, I think I have spotted a small bug here. 

The `isFullBatch` is flawed. The following scenario may happen:
```
journalSize = 1000
bufferSize = 100
queryFrom = 50 // user provided offset
queryUntil = 100 // max offset as retrieved from journalSequenceActor
```
The size of `xs` will be limited by the `queryFrom` offset. In this example, it will be 50 and it will introduce a unnecessary delay on line 180. 

I will review it in the code I sent on my last PR #127. That's probably the case there as well.

Reply:
Actually, maybe just the name if flawed, if the batchSize is equal to readJournalConfig.maxBufferSize then we can query again immediately.

if the batch size is smaller because `queryUntil` was somewhat small, then we also need to wait because the journal Sequence actor needs some time to increment the max ordering id

Reply:
> Actually, maybe just the name if flawed, if the batchSize is equal to readJournalConfig.maxBufferSize then we can query again immediately.

Maybe not, if the `queryUntil` is the last item in the batch, we may have reached the end of it, for example:
```
bufferSize = 100
queryFrom = 50 // user provided offset
queryUntil = 151 
fetchde = 51 to 151
```
But I guess this is a minor issue. Will be sorted out on next iteration.

What about that:

```scala
def decideHowToContinue = {
  val possibleRangeSize = queryUntil.maxOrdering - from
  if (possibleRangeSize > xs.size || xs.size == batchSize) {
    // there are certainly more to fetch
    Continue
  } else {
    // if query is not being limited by batch size and we fetched all possible items,
    // we can consider that there is nothing else to fetch for the moment
    ContinueDelayed
  }
}
```